# EDA Siniestros Viales

In [71]:
import re
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
import folium
from folium.plugins import HeatMap

## Extracción y lectura de los datasets

In [72]:
homicidios_hechos = pd.read_excel("data/homicidios.xlsx", sheet_name="HECHOS")
homicidios_victimas = pd.read_excel("data/homicidios.xlsx", sheet_name="VICTIMAS")

### Tratamiento de los datos del dataset `homicidios - hechos`

In [73]:
homicidios_hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH   
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4  \
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura   
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN  \
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce   
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.  \
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA   
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8  \
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y   
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022  \
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

In [74]:
homicidios_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

Examen de nulos y duplicados

In [75]:
homicidios_hechos.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

In [76]:
homicidios_hechos[["Altura", "Calle", "Dirección Normalizada", "Cruce"]][:5]

Altura             Calle   
0     NaN  PIEDRA BUENA AV.  \
1     NaN    PAZ, GRAL. AV.   
2  2034.0    ENTRE RIOS AV.   
3     NaN    LARRAZABAL AV.   
4     NaN      SAN JUAN AV.   

                               Dirección Normalizada   
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  \
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.   
2                                ENTRE RIOS AV. 2034   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   

                                 Cruce  
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.  
1                  DE LOS CORRALES AV.  
2                                  NaN  
3             VILLEGAS, CONRADO, GRAL.  
4              SAENZ PE?A, LUIS, PRES.

In [77]:
homicidios_hechos[homicidios_hechos["Dirección Normalizada"].isnull()][["Dirección Normalizada", "Cruce","Altura", "Calle"]]

Dirección Normalizada Cruce  Altura                                 Calle
38                    NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
106                   NaN   NaN     NaN     AUTOPISTA BUENOS AIRES - LA PLATA
119                   NaN   NaN     NaN                                   NaN
180                   NaN   NaN     NaN               AUTOPISTA PERITO MORENO
181                   NaN   NaN     NaN  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.
313                   NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
546                   NaN   NaN     NaN                 LUGONES, LEOPOLDO AV.
621                   NaN   NaN     NaN     AUTOPISTA BUENOS AIRES - LA PLATA

Tras examinar datos nulos, que representan el 5,1% del dataset; distribuidos en tres columnas **`Altura`**, **`Dirección Normalizada`**, **`Cruce`** y **`Calle`**. 
- Para el caso de **`Calle`** tiene un solo dato nulo.
- Para el caso de  **`Dirección Normalizada`** 8 nulos. 
- Para el caso de **`Cruce`** tiene 171 nulos, siendo el 24,56% del los datos de la columna.
- Para el caso de **`Altura`** tiene 567 nulos que representan el 81,5% de datos de la columna

Analizando los datos de la `Altura` se procede a eliminarla debido a la gran cantidad de faltantes y no existiendo manera de poder rellenar las colummas; ya que altura se refiere a la numeración de la calle donde se produce el siniestro y la mayoria de los casos se producen en cruces o esquinas.
La columna `Cruce` solo va a tener dato cuando el hecho se porduce en un cruce de calles, por lo que se conserva de esta manera.
La columna `Dirección Normalizada` no tiene posibilidad de completarse con las columnas del dataset; pero contiene solo 1,1% de nulos por lo se conserva de esta manera.

In [78]:
# Se elimina la columna "Altura"
homicidios_hechos= homicidios_hechos.drop("Altura", axis=1)
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [79]:
homicidios_hechos.duplicated().sum()

0

Para el caso no se encontraron datos duplicados 

### Transformación de los datos

Se regularizan los tipos de datos de las columnas

Ante la salida de la info del dataset, se observa que la primera columna con tipo de dato que necesita transformación es `HH`, debido a que se encuentra como objeto y deberia ser númerico.
- Por ello se analizan los datos de la columna
- Se transforman a tipo de dato numérico

In [80]:
homicidios_hechos["HH"].unique()

array([4, 1, 7, 0, 5, 18, 19, 15, 11, 22, 16, 9, 23, 6, 10, 17, 12, 8, 20,
       21, 14, 3, 2, 13, 'SD'], dtype=object)

El campo que contiene SD, no se modifica debido a que en la columna HORA donde tambien debería existir dato, se encuentra SD. Por lo que no podemos calcular un reemplazo. Otra opción sería calcular la mediana de las horas y reemplazarlas pero podria no ser representativo en este caso. Por lo que se define mantenerlo en SD

In [81]:
# Cambio el dato horas  a tipo de dato numerico
homicidios_hechos["HH"]= pd.to_numeric(homicidios_hechos ["HH"], downcast="integer", errors="coerce")

In [82]:
# Filtrar filas con posiciones no válidas
valid_positions = ~(homicidios_hechos['pos x'].str.contains('[^0-9.,-]') | homicidios_hechos['pos y'].str.contains('[^0-9.,-]') | (homicidios_hechos['pos x'] == '.') | (homicidios_hechos['pos y'] == '.'))

# Establecer posiciones no válidas en NaN
homicidios_hechos.loc[~valid_positions, ['pos x', 'pos y']] = np.nan

# Convertir las cadenas a tipo numérico (float)
homicidios_hechos['pos x'] = homicidios_hechos['pos x'].str.replace(',', '.').astype(float)
homicidios_hechos['pos y'] = homicidios_hechos['pos y'].str.replace(',', '.').astype(float)

In [83]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA   HH   
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4.0  \
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1.0   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7.0   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE             Calle   
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  PIEDRA BUENA AV.  \
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ    PAZ, GRAL. AV.   
2                         AV ENTRE RIOS 2034       AVENIDA    ENTRE RIOS AV.   

                                 Cruce   
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.  \
1                  DE LOS CORRALES AV.   
2                                  NaN   

                               Dirección Normalizada  COMUNA   
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8  \
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2                                ENTRE RIOS AV. 2034       1   

                                XY (CABA)      pos x      pos y   
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570  \
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO

La columna `XY (CABA)` posee coordenadas para poder ubicar en un mapa , las extraemos para poder utilizar

In [84]:
# Función para extraer las coordenadas CABA
# Esta función toma un argumento texto, que es el texto de entrada del cual se intentarán extraer las coordenadas
# Utiliza una expresión regular (re.findall) para buscar todas las coincidencias de números decimales (que tienen un punto como separador decimal) en el texto. Estas coincidencias se almacenan en la lista coordenadas.
# return float(coordenadas[0]), float(coordenadas[1]): Si se encontraron dos coordenadas, se convierten en números de punto flotante (float) y se devuelven como una tupla de dos valores.
# Luego, se aplica esta función a la columna 'XY (CABA)' de df_siniestros utilizando el método .map(). Esto se hace para cada fila de la columna, y los resultados se almacenan en dos nuevas columnas, 'coordenada_x' y 'coordenada_y', utilizando zip(*...).

def extraer_coordenadas(texto):
    # Utilizamos una expresión regular para extraer las coordenadas
    coordenadas = re.findall(r'\d+\.\d+', texto)
    if len(coordenadas) == 2:
        return float(coordenadas[0]), float(coordenadas[1])
    else:
        return None, None

# Aplicar la función a la columna 'XY (CABA)'
homicidios_hechos['coordenada_x'], homicidios_hechos['coordenada_y'] = zip(*homicidios_hechos['XY (CABA)'].map(extraer_coordenadas))

# Filtrar los registros que no tienen coordenadas válidas
homicidios_hechos = homicidios_hechos[(homicidios_hechos['coordenada_x'].notnull()) & (homicidios_hechos['coordenada_y'].notnull())]

In [85]:
homicidios_hechos.head(3)

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA   HH   
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4.0  \
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1.0   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7.0   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...   
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...  \
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...   
2                         AV ENTRE RIOS 2034       AVENIDA  ...   

                               Dirección Normalizada COMUNA   
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8  \
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   

                                XY (CABA)      pos x      pos y   
0   Point (98896.78238426 93532.43437792) -58.475340 -34.687570  \
1   Point (95832.05571093 95505.41641999) -58.508775 -34.669777   
2  Point (106684.29090040 99706.57687843) -58.390403 -34.631894   

    PARTICIPANTES  VICTIMA    ACUSADO   coordenada_x  coordenada_y  
0       MOTO-AUTO     MOTO       AUTO   98896.782384  93532.434378  
1  AUTO-PASAJEROS     AUTO  PASAJEROS   95832.055711  95505.416420  
2       MOTO-AUTO     MOTO       AUTO  106684.290900  99706.576878  

[3 rows x 22 columns]

In [86]:
#Se observan las columnas resultantes
homicidios_hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO', 'coordenada_x', 'coordenada_y'],
      dtype='object')

Las colummas tienen nombres en mayúsculas y también en minúsculas. Se estandarizan a modo de buena práctica

In [87]:
#Cambio la primer letra a mayúscula
homicidios_hechos.columns = [x.capitalize() for x in homicidios_hechos.columns]
# Reemplazo guiones por espacios
homicidios_hechos.columns = homicidios_hechos.columns.str.replace("_", " ")
#Renombra columnas 
homicidios_hechos= homicidios_hechos.rename(columns={"N victimas": "Num víctimas",
                                            "Aaaa": "Año",
                                            "Mm":"Mes",
                                            "Dd": "Día",
                                            "Hh": "Hora int"}) 

homicidios_hechos.columns

Index(['Id', 'Num víctimas', 'Fecha', 'Año', 'Mes', 'Día', 'Hora', 'Hora int',
       'Lugar del hecho', 'Tipo de calle', 'Calle', 'Cruce',
       'Dirección normalizada', 'Comuna', 'Xy (caba)', 'Pos x', 'Pos y',
       'Participantes', 'Victima', 'Acusado', 'Coordenada x', 'Coordenada y'],
      dtype='object')

### Tratamiento de los datos del dataset `homicidios- víctimas`

In [88]:
homicidios_victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


In [89]:
homicidios_victimas.isnull().sum()

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64

### Tratamiento de los datos del dataset `lesiones`